processing/src/main/java/io/druid/query/groupby/epinephelinae/RowBasedGrouperHelper.java

```diff
@@ -918,19 +970,59 @@ private static int compareDimsInRowsWithAggs(
         final List<DimensionSpec> dimensions,
         final long maxDictionarySize,
         final DefaultLimitSpec limitSpec,
-        final List<ValueType> valueTypes
+        final List<ValueType> valueTypes,
+        @Nullable final List<String> dictionary
     )
     {
       this.includeTimestamp = includeTimestamp;
       this.sortByDimsFirst = sortByDimsFirst;
       this.dimensions = dimensions;
       this.dimCount = dimensions.size();
-      this.maxDictionarySize = maxDictionarySize;
       this.valueTypes = valueTypes;
       this.limitSpec = limitSpec;
-      this.serdeHelpers = makeSerdeHelpers();
+      this.enableRuntimeDictionaryGeneration = dictionary == null;
+      this.dictionary = enableRuntimeDictionaryGeneration ? new ArrayList<>(DICTIONARY_INITIAL_CAPACITY) : dictionary;
+      this.reverseDictionary = enableRuntimeDictionaryGeneration ?
+                               new Object2IntOpenHashMap<>(DICTIONARY_INITIAL_CAPACITY) :
+                               new Object2IntOpenHashMap<>(dictionary.size());
+      this.reverseDictionary.defaultReturnValue(UNKNOWN_DICTIONARY_KEY);
+      this.maxDictionarySize = maxDictionarySize;
+      this.serdeHelpers = makeSerdeHelpers(limitSpec != null, enableRuntimeDictionaryGeneration);
```

Comment:
Consider converting to array, for faster access

Reply:
Hmm, does it give us a noticeable benefit? I'm not sure because serdeHelpers will be created per dimension and the number of dimensions won't be extremely large.

Reply:
Effect probably won't be big, but it also costs nothing to make this change. I think it's a good rule of thumb, when preparing some small iterables, which will be iterated much more times later, to store them as arrays rather than array lists.

Reply:
Agree. I asked you because you said it'll be faster. Changed.